components/controller_host.py

```diff
@@ -0,0 +1,139 @@
+from qunetsim.components import Host
+from utils import DefaultOperationTime
+
+
+class ControllerHost(Host):
+    """
+    Controller host object which acts as a master node in a centralised
+    distributed network system.
+    """
+
+    def __init__(self, host_id, computing_host_ids, gate_time=None):
+        """
+        Returns the important things for the controller hosts
+
+        Args:
+            host_id (str): The ID of the controller host
+            computing_host_ids (list): The IDs of computing/slave hosts
+            gate_time (dict): A mapping of gate names to time the gate takes to
+               execute for each computing host
+        """
+        super().__init__(host_id)
+
+        self._computing_host_ids = computing_host_ids
+        self.add_c_connections(computing_host_ids)
+
+        if gate_time is None:
+            gate_time = {}
+            for computing_host_id in computing_host_ids:
+                gate_time[computing_host_id] = DefaultOperationTime
+
+        self._gate_time = gate_time
+
+    @property
+    def computing_host_ids(self):
+        """
+        Get the *computing_host_ids* associated to the controller host
+        Returns:
+            (list): The IDs of computing/slave hosts
+        """
+        return self._computing_host_ids
+
+    def add_computing_host_to_network(self, computing_host_id, gate_time=None):
+        """
+        Adds a computing host to the distributed network
+
+        Args:
+            computing_host_id (str): The ID of the computing host
+            gate_time (dict): A mapping of gate names to time the gate takes to
+               execute for the computing host to be added
+        """
+
+        self._computing_host_ids.append(computing_host_id)
+        self.add_c_connection(computing_host_id)
+
+        if gate_time is None:
+            gate_time = DefaultOperationTime
+
+        self._gate_time[computing_host_id] = gate_time
+
+    def add_computing_hosts_to_network(self, computing_host_ids, gate_time=None):
+        """
+        Adds multiple computing hosts to the distributed network
+
+        Args:
+            computing_host_id (List): The ID of the computing host
+        """
+
+        for computing_host_id in computing_host_id:
+            self._computing_host_ids.append(computing_host_id)
+            self.add_c_connection(computing_host_id)
+
+            if gate_time is None:
+                gate_time = default_gate_time
+
+            self._gate_time[computing_host_id] = gate_time
+
+    def create_distributed_schedule(self, circuit):
```

Comment:
I guess this can be private i.e. `_create_distributed_schedule`... and then something like `generate_schedules` to expose it which also sends the schedules to the computing hosts...